src/vector_windows.rs

```diff
@@ -121,6 +167,41 @@ pub mod service_control {
         Ok(())
     }
 
+    fn restart_service(service_def: &ServiceDefinition) -> std::result::Result<(), Error> {
+        let service_access =
+            ServiceAccess::QUERY_STATUS | ServiceAccess::START | ServiceAccess::STOP;
+        let service = open_service(&service_def, service_access)?;
+        let service_status = service.query_status()?;
+
+        if service_status.current_state == ServiceState::StartPending
+            || service_status.current_state == ServiceState::Running
+        {
+            service.stop()?;
```

Comment:
So this is doing like an `HUP` on Linux? I know on Linux we try to ensure the API server etc don't need to restart when we get this signal.

Reply:
I don't think that it maps to a particular signal on Linux. Windows does not have a notion of signal.

It actually calls the winsvc [ControlService](https://docs.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-controlservice) function with a `SERVICE_CONTROL_STOP` control code, which will then send a stop notification to the service (the event handler of the service will be notified with a `ServiceControlStop`.

It won't restart when you only send a Stop notification.

Reply:
So does this mean the API server would go down during the restart? I know @leebenson has been fairly careful to make sure the API server stands persisting over restarts.

Reply:
Oh ok, I see what you mean.

Yes, the API server will also restart when issuing a restart. What you describe would be a reload then (the `SignalTo::Reload`).

Windows has a `SERVICE_CONTROL_PARAMCHANGE` control code that we could use to trigger a reload. I would personnally keep the behavior of `service restart` as-is, and maybe add a `service reload` sub command which would then use `SERVICE_CONTROL_PARAMCHANGE` to instruct the service to reload its parameters.

What do you think ?

Reply:
I'm not especially familiar with Windows, but if `service restarts` is intended to mean 'hard stop, then start', I think it's fine to also bring down the API. This should only happen where the user wants to tear down the entire Vector process, and start a new one in its place.

`service reload` sounds equivalent to `HUP`; in the situation where a config is being reloaded and topology needs to tear down and start again, the API should remain up. It's been specifically designed to allow connecting API clients to remain active while hot config changes are loaded. This facilitates our eventual work with a web GUI that will allow users to make live config changes, validate config, and see toplogy graphs respond to updates in real-time.

Does that fit with your model here, @oktal?

Reply:
Yup, that sounds good to me.

`service restart` -> triggers a hard stop, then start (you might want to do that upon version update, and not only config change)
`service reload` -> only trigger a reload of the running topology, without bringing down the entire service

I would keep the changes separate and implement `reload` on a separate `windows-service-reload` branch if that sounds good to you @Hoverbear :)

Reply:
Sounds great! Let's get this merged then!